### PR TITLE
Increase safety of terraform provider by adding `nilaway` | Merge taskfiles

### PR DESCRIPTION
**Pull request by taimoorgit**
## Issues

https://github.com/OpsLevel/team-platform/issues/186

## Changelog

- [x] Make changes for nilaway to pass
- [x] Add nilaway to task file
- [x] Merge taskfiles into one, improve ordering/structure/aliases
- [x] Added the standard expected tasks, like `task ci` and `task fix`
- [ ] ~~Make a `changie` entry~~

## Tophatting

All `task` commands worked fine in root directory but not in opslevel/ [because special variables in Taskfile](https://taskfile.dev/api/#special-variables) don't work that well.

```
$ task
task: Available tasks for this project:
* ci:                      Workflow to run in CI
* fix:                     Fix formatting, linting, go.mod, and update submodule
* lint:                    Formatting and linting
* setup:                   Setup linter, formatter, etc. for local testing
* setup-terraform:         Setup env and tools                                                                     (aliases: setup-tf, tf-setup, terraform-setup)
* terraform-apply:         Rebuild and run "terraform apply -auto-approve" in "workspace"                          (aliases: apply)
* terraform-build:         Build local opslevel terraform provider                                                 (aliases: build)
* terraform-clean:         Completely wipe terraform state and all terraform generated files from "workspace"      (aliases: clean)
* terraform-destroy:       Rebuild and run "terraform destroy -auto-approve" in "workspace"                        (aliases: destroy)
* terraform-init:          Initialize terraform workspace                                                          (aliases: init)
* terraform-plan:          Rebuild and run "terraform plan" in "workspace"                                         (aliases: plan)
* test:                    Run tests
* workspace:               Setup workspace for terraform-provider-opslevel & opslevel-go development
$ task ci && task fix && task lint && task setup && task terraform-plan && task test && task workspace
# no errors
```

### CODE REVIEW

File: Taskfile.yml [hunk 1]
```diff
@@ -2,28 +2,92 @@
...
-  # 'VERSION_DYNAMIC' needs to run in this scope before 'task build' begins
+  # 'VERSION_DYNAMIC' needs to run in this scope before 'task terraform-build' begins
   VERSION_DYNAMIC:
     sh: echo "$(date +%s).0.0"
   WORKSPACE_DIR: "workspace"
 
-includes:
-  go:
-    internal: true
-    taskfile: "./opslevel/Taskfile.yml"
```
Comment (davidbloss) [inline Taskfile.yml:14]: What is the motivation / benefit behind moving the contents of the opslevel (Go specific) Taskfile into this root level Taskfile?

File: opslevel/datasource_opslevel_filter.go [hunk 2]
```diff
@@ -54,10 +54,13 @@ func filterFilters(data []opslevel.Filter, field string, value string) (*opsleve
...
 func datasourceFilterRead(d *schema.ResourceData, client *opslevel.Client) error {
 	resp, err := client.ListFilters(nil)
-	results := resp.Nodes
 	if err != nil {
 		return err
 	}
+	if resp == nil {
+		return fmt.Errorf("unexpected: listing filters returned nil")
+	}
+	results := resp.Nodes
```
Comment (rocktavious) [inline opslevel/datasource_opslevel_filter.go:63]: what about if `resp.Nodes` is empty?

File: opslevel/resource_opslevel_check_repository_search.go [hunk 3]
```diff
@@ -35,7 +37,11 @@ func resourceCheckRepositorySearchCreate(d *schema.ResourceData, client *opsleve
 	checkCreateInput := getCheckCreateInputFrom(d)
 	input := opslevel.NewCheckCreateInputTypeOf[opslevel.CheckRepositorySearchCreateInput](checkCreateInput)
 	input.FileExtensions = opslevel.RefOf(getStringArray(d, "file_extensions"))
-	input.FileContentsPredicate = *expandPredicate(d, "file_contents_predicate")
+	predicate := expandPredicate(d, "file_contents_predicate")
+	if predicate == nil {
+		return fmt.Errorf("unexpected: expanding predicate 'file_contents_predicate' returned nil")
```
Comment (rocktavious) [inline opslevel/resource_opslevel_check_repository_search.go:42]: think about this error message for a second.  This is what the end user is going to read - they have no idea what `expanding` means and they have no idea why it might be `nil` especially if they've actually configured it.  We should write a better / more helpful error message.  What are the error cases with `expandPredicate` (if any) - can we catch each one inside `expandPredicate` and provide specific helpful information for that error case to the end user so they know exactly what they need to fix?

File: opslevel/resource_opslevel_service.go [hunk 4]
```diff
@@ -165,8 +150,12 @@ func reconcileTags(d *schema.ResourceData, service *opslevel.Service, client *op
 	}
 	tagInput := map[string]string{}
 	for _, tag := range tags {
-		key := strings.TrimSpace(strings.Split(tag, ":")[0])
-		value := strings.TrimSpace(strings.Split(tag, ":")[1])
+		parts := strings.Split(tag, ":")
+		if len(parts) != 2 {
+			return fmt.Errorf("invalid tag: '%s'", tag)
```
Comment (rocktavious) [inline opslevel/resource_opslevel_service.go:155]: lets fix this error message too - give them the exact reason why the tag is invalid - IE tag should have a very specific format with `:` etc - much like you did for the id error in property assignment

File: opslevel/datasource_opslevel_filters.go [hunk 5]
```diff
@@ -26,10 +28,13 @@ func datasourceFilters() *schema.Resource {
 
 func datasourceFiltersRead(d *schema.ResourceData, client *opslevel.Client) error {
 	resp, err := client.ListFilters(nil)
-	result := resp.Nodes
 	if err != nil {
 		return err
 	}
+	if resp == nil {
+		return fmt.Errorf("unexpected: listing filters returned nil")
```
Comment (rocktavious) [inline opslevel/datasource_opslevel_filters.go:35]: is this truly a problem to halt the execution of terraform plan?  Couldn't we defensively code this such that is the request to ListFilters returned an empty array that we store into the state an empty array too?

Additionally is it ever the case where `resp` == nil and `err` != nil ?

**Review by rocktavious — CHANGES_REQUESTED** [review]
A boatload of questions and some better user facing error messages are needed

File: Taskfile.yml (hunk 1 above)
Comment (taimoorgit) [reply Taskfile.yml:14]: @davidbloss Overall I do not see the benefit in having 2 Taskfiles. I understand that the benefits of having two Taskfiles is that we get a separation of responsibilities where the Go Taskfile handles Go-specific things only.

However I do not think that that is enough of a benefit compared to the drawbacks.

1. It's harder to keep the Taskfile and Go-specific Taskfile in sync. If you delete or rename a task from file A it's not easy to remember to update the other Taskfile. I understand that you can use refactoring tools and plugins in an IDE to do this, but that adds overhead to something that should be simple.
2. It's harder to read the Taskfile's order of execution and to figure out what's doing what because you have to jump between multiple files. 
3. All of our other tools that are using Taskfiles are using a single file. This makes this repo's Taskfile the odd one out, and makes keeping it in sync with Taskfiles from other even repos harder.
4. Adding to this, it would be really annoying if each repo started having its own nested Taskfiles that are slightly different from each other. Taskfiles are not long and complex enough to warrant them not all being nearly identical.
5. I would prefer that tasks are differentiated by their name (e.g. terraform stuff is prefixed with `terraform-`) rather than what file they are contained in. It is much easier to understand what `task terraform-apply` does from its name compared to `task apply`.
6. The separation of responsibility isn't so clear. It made sense when we had Taskfile 1 that had all the basic commands (like `ci, setup, lint, fix, workspace`) and then Taskfile 2 for terraform commands and Taskfile 3 for Go commands. But now Taskfile 1 is doing basic commands AND terraform whereas Taskfile 2 is specific to only one thing. If we added another dependency that doesn't relate to Go/Terraform, should it be placed into its own separate Taskfile, or should it go into Taskfile 1? By placing it into its own Taskfile it would make our Taskfile hierarchy even more painful to maintain, which is why we removed the terraform Taskfile already. By placing it into the root Taskfile, the root Taskfile would seem crowded in comparison compared to the Go Taskfile since it's handling 3 things at once.

TL;DR I value uniformity of Taskfiles more than trying to separate them by responsibility.

File: opslevel/datasource_opslevel_filter.go (hunk 2 above)
Comment (taimoorgit) [reply opslevel/datasource_opslevel_filter.go:63]: @rocktavious `resp.Nodes` here is only being used for `range` which is why `nilaway` only warned me about line 57 in the old code. Therefore this nil check is sufficient. [(link to usage of `resp.Nodes`)](https://github.com/OpsLevel/terraform-provider-opslevel/blob/920c247641f67d1fdcf431d9f2cc4bbf3b98239a/opslevel/datasource_opslevel_filter.go#L31)

In GoLang arrays and maps are interfaces and doing `range` or even `len` over them will not cause a nil pointer panic. [Demonstration](https://go.dev/play/p/_GWLMuyxcLh)

If someone does make an update to do something than other `range` over the `resp.Nodes`, it will trigger `nilaway` and force them to add nil checks.

File: opslevel/datasource_opslevel_filters.go (hunk 5 above)
Comment (taimoorgit) [reply opslevel/datasource_opslevel_filters.go:35]: >Couldn't we defensively code this such that is the request to ListFilters returned an empty array that we store into the state an empty array too?

We could, but in that case I think it would make more sense for `ListFilters` to not return a pointer, but rather return a slice. I have thought about the need to add this change for a while, since it is of 0 value for client applications to receive `WhateverConnection` types since they're essentially only useful for graphql requests whereas the consumers of opslevel-go are only concerned about what the actual data is. That would be out of scope for this PR. 

Right now the behaviour of terraform is to just crash if it encounters `resp == nil`, so this check is still better than what we had before. In the case that `resp != nil` and `len(resp.Nodes) == 0` the behaviour is completely unchanged. 

>Additionally is it ever the case where resp == nil and err != nil ?

With the opslevel-go code as it is now, this will never happen AFAICT. But there is no guarantee as to what opslevel-go will return as far as the terraform provider knows. It is totally possible that in the future we will accidentally make a change that results in `resp == nil && err != nil` and that would cause the terraform client to panic. 

`nilaway` supports annotations that we can add to opslevel-go to prevent it from alerting on this, but that would also not prevent the terraform provider from crashing on an erroneous change which would reduce how defensive our code is.

File: opslevel/resource_opslevel_check_repository_search.go (hunk 3 above)
Comment (taimoorgit) [reply opslevel/resource_opslevel_check_repository_search.go]: @rocktavious you are correct, this error message makes no sense and would only frustrate end users. I introduced it because of a change I was making and reverted but this was leftover.

I am fixing this to instead return an error explaining that a valid `file_contents_predicate` was expected to be found in the configuration for this request, but there was none found. 

Changed error message to be:

```
		return fmt.Errorf("error reading 'file_contents_predicate' in config - a valid type and value are required")
```

which matches the schema that `expandPredicate()` tries to parse from a config.

File: opslevel/resource_opslevel_service.go (hunk 4 above)
Comment (taimoorgit) [reply opslevel/resource_opslevel_service.go]: Updated the error message to be more descriptive, just like when reading terraform property assignment IDs.

```
return fmt.Errorf("[%s] invalid tag, should be in format 'key:value' (only a single colon between the key and value, no spaces or special characters)", tag)
```

File: opslevel/datasource_opslevel_filters.go (hunk 5 above)
Comment (rocktavious) [reply opslevel/datasource_opslevel_filters.go:35]: ```since it is of 0 value for client applications to receive WhateverConnection types since they're essentially only useful for graphql requests```

is not true - List calls are recursive so the connection object is needed to properly perform pagination

Comment (taimoorgit) [reply opslevel/datasource_opslevel_filters.go:35]: > `since it is of 0 value for client applications to receive WhateverConnection types since they're essentially only useful for graphql requests`
> 
> is not true - List calls are recursive so the connection object is needed to properly perform pagination

OK, after re-thinking this it might be useful in certain applications that consume opslevel-go to retain the original pagination information that is given by a `Connection`. For example a use case would be to search in the first N pages of results but stop after that. There is also the benefit of `WhateverConnection` objects holding things like `[]ServiceId` instead of a massive `[]Service`. However I don't think any of our existing tools (CLI/TF/K8S) are taking advantage of this. 

In that case my idea of:

> it would make more sense for ListFilters to not return a pointer, but rather return a slice.

would not be useful.

I still think that the `nil` check should be retained however. I agree that it is bothersome, but we should either

1. Retain it, since a pointer is returned at the end of the day
2. Refactor `ListWhatever()` functions to always return `(ThingConnection, error)` instead of `(*ThingConnection, error)`.

Comment (rocktavious) [reply opslevel/datasource_opslevel_filters.go:35]: Ok lets leave the nil check for now then.

File: Taskfile.yml (hunk 1 above)
Comment (davidbloss) [reply Taskfile.yml:14]: These Taskfiles are supposed to be a utility to make developing easier, not to become the an end in themselves.

On the drawbacks:

1. This surprises me, we're working with two files here and this sounds fairly hypothetical. Code reviews help to catch any rename/delete misses. If this truly stressful work I would be happy to take Taskfile related tickets in the future. Plugins and refactoring tools sound like overkill for this kind of work. I usually opt for a vertically split window. Also, take a look at Gitlab's web of [nested yml files](https://github.com/gitlabhq/gitlabhq/blob/master/.gitlab/ci/rails.gitlab-ci.yml) - it could be much worse.
2. IMO it's actually harder to scroll up and down searching through an extra 100 lines of code. Viewing two distant parts of a long file, or two parts of two different files can be solved by a split window.
3. True, the other repos have a single Taskfile, but all of our other tools are purely Go code bases. For this repo with Go and Terraform code, we're primarily interested in configuring/running Terraform operations and need to build the Go code as a necessary step. This repo is the odd one out by design and the Taskfile's match that. 
4. I didn't think our code bases were big enough for this concern. 🤷‍♂️ 
5. 👍 Certainly, whatever makes our work easier. I see you added aliases to those tasks.
6. I think this should have been a meeting or at least a discussion earlier on. I had no idea this was causing such a headache. The terraform Taskfile was buggy due to the working directory that we were running tasks in. Before, `task plan` just ran `terraform plan`. With `terraform plan` we have to be in the right directory, with `task plan` the idea was we could run `terraform plan` from any directory. Consolidating the terraform Taskfile into the root Taskfile was to get around all the directory changing and careful use of these [special variables](https://taskfile.dev/api/#special-variables).

I understand that we all have preferences but this is a pretty hefty change for essentially the same command line UX and feature set. Please lmk if there is anything else like this that's impeding your work.

Comment (taimoorgit) [reply Taskfile.yml:14]: @davidbloss fair point regarding the nested yml files. I am going to revert the merge but keep the new aliases and tasks.

Comment (taimoorgit) [reply Taskfile.yml:14]: @davidbloss I unmerged everything now there's two taskfiles again but I still require use of the special variables.

Comment (taimoorgit) [reply Taskfile.yml:14]: resolving conversation - 1 taskfile is the way to go because controlling directories is very difficult AND there isn't really a reason to run a taskfile in the opslevel/ dir and not the root dir.

File: Taskfile.yml [hunk 6]
```diff
@@ -3,27 +3,79 @@
...
+    desc: Workflow to run in CI
+    deps:
+      - brew-install-jq
+      - install-gofumpt
+      - install-golangci-lint
+      - install-nilaway
+    cmds:
+      - task: workspace
+      - task: has-latest-opslevel-go
+      - task: lint
```
Comment (davidbloss) [inline Taskfile.yml:23]: Small but worth pointing out. Let's move `task: lint` before `task: workspace`. (I think this update was made to `kubectl-opslevel` since that was taking a long time)

Putting `lint` first will ensure we're only lint checking the code in this repo, and not the submodule repos also. Plus it's a little bit faster.

File: Taskfile.yml [hunk 7]
```diff
@@ -34,108 +86,135 @@ tasks:
...
-      - task: go:test
+      - go test -race -coverprofile=coverage.txt -covermode=atomic -v ./... {{ .CLI_ARGS }}
+
+  workspace:
+    desc: Setup workspace for terraform-provider-opslevel & opslevel-go development
+    cmds:
+      - cmd: echo "Setting up opslevel-go workspace..."
+        silent: true
+      - git submodule update --init --remote
+      - rm go.work go.work.sum || true
```
Comment (davidbloss) [inline Taskfile.yml:137]: 🚀 good move anyway and hopefully resolves some sneaky local dev issues 🙌

**Review by davidbloss — APPROVED** [review]
I left one non-blocking comment in running `lint` earlier. But overall looks good! 🎉

File: Taskfile.yml (hunk 6 above)
Comment (taimoorgit) [reply Taskfile.yml]: done

Comment (taimoorgit) [reply Taskfile.yml:23]: after moving it before workspace CI failed because of unresolved references

https://github.com/OpsLevel/terraform-provider-opslevel/actions/runs/7762742933/job/21173702092

**Review by rocktavious — APPROVED** [review]
